### ISSUE TITLE
No working Never typing in BaseClass

### ISSUE THREAD

**Issue by ArtemIsmagilov** [pydantic/pydantic#9731]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

I get an error

```python
pydantic.errors.PydanticSchemaGenerationError: Unable to generate pydantic-core schema for typing.Never. Set `arbitrary_types_allowed=True` in the model_config to ignore this error or implement `__get_pydantic_core_schema__` on your type to fully support it.

If you got this error by calling handler(<some type>) within `__get_pydantic_core_schema__` then you likely need to call `handler.generate_schema(<some type>)` since we do not call `__get_pydantic_core_schema__` on `<some type>` otherwise to avoid infinite recursion.

For further information visit https://errors.pydantic.dev/2.7/u/schema-for-unknown-type
```

### Example Code

```Python
class PersonFilmRoles(BaseModel):
    id: str
    roles: list[str | Never]
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.7.4
pydantic-core version: 2.18.4
python version: 3.12.3
```

**Comment by sydney-runkle** [comment]
Hi @ArtemIsmagilov,

Thanks for reporting this. I'd consider this more of a feature request - I can't really understand a practical example of where you'd want to use the `typing.Never` hint for a model field - could you shed some light on that?

If there's not a practical example, I think we'll probably mark this as not planned.

**Comment by ArtemIsmagilov** [comment]
I want to specify a model that would accept only an empty list, without elements. to do this, I'm creating the following class. For example, I can use this validator for the API

newbie actor at first job
```python
class PersonFilmRoles(BaseModel):
    id: str
    roles: list[Never]

newbie_actor = PersonFilmRoles(id='id', roles=[])
```

**Comment by sydney-runkle** [comment]
Hmm,

For this case, I'm still not convinced, because I feel as though

```py
from typing import Annotated

from pydantic import BaseModel
from annotated_types import MaxLen


class PersonFilmRoles(BaseModel):
    id: str
    roles: Annotated[list, MaxLen(0)]


newbie_actor = PersonFilmRoles(id='id', roles=[])
```

Would be more appropriate.

That being said, thinking about it more, perhaps it makes sense as an attribute annotation if you seek to distinguish a more abstract base class from a concrete one, so I'll leave this as a feature request.

Feel free to open a PR adding support!

**Comment by ArtemIsmagilov** [comment]
To be honest, I don't understand you a little. The standard type doesn't work for you. Is this really not a bug but a feature?

**Comment by ArtemIsmagilov** [comment]
> Feel free to open a PR adding support!
Thanks)

**Comment by sydney-runkle** [comment]
I suppose the categorization here isn't super critical - it's low priority either way, but we welcome PRs from the community :)

**Comment by Viicos** [comment]
I'm not sure I really understand the use case here. `Never` is the bottom type, it has a [specific meaning in the typing spec](https://typing.readthedocs.io/en/latest/spec/special-types.html#never). Using it to indicate that a list has no element does not really follow its semantics. At the very least, I think it could make sense to have:

```python
from typing import Never

class Model(BaseModel):
    foo: Never

Model(foo=...)  # Always fails, no matter the giving value of `foo`
```

**Comment by davidhewitt** [comment]
Xref https://github.com/pydantic/pydantic-core/issues/619

**Comment by scullion** [comment]
To offer a practical use case, I've found `Never` handy in "knocking out" a field from a compound type for the purposes of writing `@overload` signatures. When a field's type comes from a type parameter, instantiating the containing structure with that parameter set to `Never` creates a type in which that field must be absent (or, as in the example below, must be `None`).

```python
H = TypeVar('H', bound=DriveCore)

class HyperdriveConfig(BaseModel, Generic[H]):
    core: H | None = None
    
    antimatter_containment_pressure: float = 8.1e41
    flux_capacitance_ratio: float = 13.41
    graviton_polarization_angle: float = 0.0

    # ... 106 other fields ...

    def recalibrate(self, updates: HyperdriveConfig[Never]) -> Self:
        return self.model_copy(update=updates.model_dump())


drive = HyperdriveConfig(core=StandardCore(), flux_capacitance_ratio=-1.8)

# these are fine because they don't replace the core with an incompatible type
drive.recalibrate(HyperdriveConfig(antimatter_containment_pressure=1e42)) 
drive.recalibrate(HyperdriveConfig(core=None)) 

# ...but thanks to Never, this will not type check.
drive.recalibrate(HyperdriveConfig(core=DangerouslyIncompatibleCore())) 
```

The above will (presumably) cause a runtime error from Pydantic, but it would be nice if this worked. I landed here after trying to do something similar in a real project, and would make use of it if it were supported.

**Comment by Viicos** [comment]
Hum, I'm not sure `Never` fulfills your use case. If you only want `HyperdriveConfig` instances without any `core` set, the signature should look like `recalibrate(self, updates: HyperdriveConfig[None]) -> Self`.

**Comment by scullion** [comment]
> If you only want `HyperdriveConfig` instances without any `core` set, the signature should look like `recalibrate(self, updates: HyperdriveConfig[None]) -> Self`.

That wouldn't type check because `None` is not compatible with the type bound on `H`. The bound would have to be relaxed to include `None`.

```python
H = TypeVar('H', bound=DriveCore | None)
```

But that necessitates further complications if `H` is used in other places where `None` must be prohibited: now a second `TypeVar` must be introduced.

And even that doesn't help if `None` is not a permissible value for the field.

```python
core: H = Field(default_factory=IdleCore)
```

In that case a new type, compatible with the bound, has to be invented as a substitute for `Never`.

```python
class NeverCore(DriveCore): ...
ConfigWithoutCore = HyperdriveConfig[NeverCore]
```

This is the workaround used in the code that inspired this example.

**Comment by sydney-runkle** [comment]
I'd like to try this for v2.10 :).

**Comment by Viicos** [comment]
@scullion I see. But your use case might be a bit different: `Never` is the bottom type, meaning `Never | <tp>` ~ `<tp>`. In your example:
- from a type checking perspective (i.e. during static analysis), `HyperdriveConfig[Never]` means any instance of `HyperdriveConfig` where `core` is `None` (and only `None`) can be used, because `Never | None` ~ `None`.
- from a Pydantic perspective (i.e. runtime analysis), `HyperdriveConfig[Never]` will create a new Pydantic with a new core schema generated. During the creation of this new model, type variables will be "replaced" by the provided type, but no simplification of the type is performed. Pydantic will thus try to generate a schema for the field `core` of type `Never | None`, and will fail because it doesn't know about `Never`.

Pydantic does not currently do any simplification of types because generally the user can do it when defining a model (i.e. there's no reason to define a field with type `Any | None`, just use `Any`). However, it could make sense when parametrizing generics. I think it is reasonable to add support for `Any | <tp>` (~ `Any`) and `Never | <tp>` (~ `<tp>`). This would solve your issue without having to add support for `Never` yet.

**Comment by scullion** [comment]
There seems to be some resistance to supporting `Never` 🙂 If that comes from a desire not to add code that "should never need to be run", I appreciate that, and can only say that the always-fails validator proposed in https://github.com/pydantic/pydantic-core/issues/619 would serve the same pupose as error handling code a Pydantic user is currently obliged to write if they define a dummy "never type" to work around the lack of `Never` support, such as the `NeverCore` constructor in the example below.

What @Viicos propses solves my real life problem, but doesn't handle the case where the type var isn't used as part of a union as shown below. Currently if the user has a field typed `T` rather than `T | None`, they are obliged to invent a type compatible with `T` like `NeverCore` to achive an effect similar to `Never`.

```python
class NeverCore(DriveCore):
    def __init__(self):
        raise RuntimeError("something is quite seriously wrong")

class HyperdriveConfig(BaseModel, Generic[H]):
    core: H = Field(default_factory=IdleCore)

    def recalibrate(self, updates: HyperdriveConfig[NeverCore]) -> Self:
        return self.model_copy(update=updates.model_dump())
    
    def install_core(self, core: H) -> HyperdriveConfig[H]:
        self.core.safe_power_down()
        return self.model_copy(update={'core': core})
    
# it would be dangerous to replace the core during a recalibration
drive.recalibrate(HyperdriveConfig(core=FluxGravitonCore())) # type error

# ... instead should to use install_core()
drive.install_core(FluxGravitonCore()) # ok
```

Isn't having Pydantic do type algebra on unions harder to implement than just having an always-fails validator? Either would solve my use case, but as a user I'd prefer first class support for `Never` purely on the basis of the principle of least surprise.

**Comment by Viicos** [comment]
> There seems to be some resistance to supporting `Never`

Not at all! I'm just stating that your use case can be treated differently. To be clear, adding support for `Never` will also solve your use case. `Never | None` will be supported when building the schema, and validation will work as expected. That is, for a given value `v`:
- `v` will try to be validated against `Never`, and fails because `Never` should always fail validation.
- `v` will try to be validated against `None`, and will fails if `v is not None`.

But having "type algebra on unions" just for this case just makes things cleaner imo. The resulting core schema is smaller and probably easier to understand. The generated JSON Schema might be simplified as well.

**Comment by sydney-runkle** [comment]
This would be a cool feature for sure. I've bumped this to v2.11 - PRs welcome with some help here! Not quite high priority enough to fit into our current sprint.

**Comment by sydney-runkle** [comment]
This is lower priority than I thought - bumped out of milestone. PRs still welcome, of course.

**Comment by Viicos** [comment]
@ArtemIsmagilov, note that if this is going to be implemented, it will be following the semantics specified in https://github.com/pydantic/pydantic/issues/9731#issuecomment-2194319078 (just wanted to make sure you were aware of it).

**Comment by iliya-malecki** [comment]
Hi y'all! I have, potentially, another need for `Never` support. In SQLModel, im representing a database type that has no meaning in python, and must never be accessed - neither selected, nor inserted. I would love to represent it as:
```python
from sqlmodel import SQLModel, Field
from sqlalchemy.types import UserDefinedType

class Earth(UserDefinedType):
    def get_col_spec(self, **kwargs):
        return "earth"

class MyTable(SQLModel, table=True):
    earth: Never = Field(sa_type=Earth)
```
The idea is that while it makes sense to index it or run sql-specific queries on it, it has no meaning ouside of its own earthdistance extension

**Comment by CyberTailor** [comment]
Another use case for `typing.Never`:

```python
class FooDict(TypedDict)
    # API returns either a dict or an empty list
    x: dict[str, str] | list[Never]

foo: FooDict
for result in dict(foo.get("x", {})).values():
    ...
```

**Comment by CyberTailor** [comment]
My workaround:

```python
# SPDX-License-Identifier: CC-BY-SA-4.0
from typing import Annotated, Never
from pydantic import Field, GetPydanticSchema

HandleAsNone = GetPydanticSchema(lambda _s, h: h(None))
type EmptyList = Annotated[list[Annotated[Never, HandleAsNone]],
                           Field(max_length=0)]
```

**Comment by Chaz6** [comment]
Another use case for this is with OCI Container Images, specifically the configuration [1] of volumes.

```
Volumes object, OPTIONAL

A set of directories describing where the process is likely to write
data specific to a container instance. NOTE: This JSON structure
value is unusual because it is a direct JSON serialization of the Go
type map[string]struct{} and is represented in JSON as an object
mapping its keys to an empty object.
```

This could be represented as `dict[Never, Never]`. We can protect this at runtime like so:

```python
from typing import Annotated, TypeAlias
from pydantic import BaseModel, AfterValidator, Field

def check_empty(v: dict) -> dict:
    if len(v) != 0:
        raise ValueError("Dictionary must be empty")
    return v

EmptyDict: TypeAlias = Annotated[
    dict, 
    AfterValidator(check_empty),
    Field(default_factory=dict)
]

class MyModel(BaseModel):
    empty_dict: EmptyDict
```

[1] https://specs.opencontainers.org/image-spec/config/

**Comment by Viicos** [comment]
Same as https://github.com/pydantic/pydantic/issues/9731#issuecomment-2194319078, `Never` isn't really meant for this. You should use `Annotated[dict[Any, Any], Field(max_length=0)]` instead.

**Comment by Youssefares** [comment]
Hey team, haven't contributed for a while, this seemed like a good first issue back/good to have. Can see now I need the issue assigned to me before opening a PR? Is this deferred for a specific reason or just low priority? 

If you're happy for me to work on this, please assign to me :) Have took on the feedback from the first iteration in #1579 and have a 3 file PR implementing a pydantic-only approach as recommended.

**Comment by Viicos** [comment]
Hi @Youssefares, regarding this issue, I feel like we should hold off on this issue until we find a real use case to have a type that always fail validation (as you can see, there is some confusion from users in this thread). If I find a good issue to contribute, I'll make sure to let you know!

**Comment by Youssefares** [comment]
Np! For reference, this sort of "force a value to be absent in a generic subtype" setup that I have in my tests and that was mentioned in this issue is the only example I could think of. It's of course achievable in a slightly different way by just doing Generic[None].

There's a **legitimate** case where Never is the only valid option and no workaround exists: _bounded TypeVars._

```python
from typing import TypeVar, Generic
from typing_extensions import Never
from pydantic import BaseModel

T = TypeVar('T', bound=int)

class NumericContainer(BaseModel, Generic[T]):
    value: T | None = None
    label: str = ''
```

If you want a variant where `value` can only be `None`, the obvious approach is `NumericContainer[None]` but mypy rejects it:

```
error: Type argument "None" of "NumericContainer" must be a subtype of "int"  [type-var]
```

`Never` is the bottom type — a subtype of everything, including `int` so `NumericContainer[Never]` passes type checking. The field `T | None` collapses to `Never | None` which is just `None`. Without pydantic supporting `Never`, the runtime crashes even though the type checker says it's valid.

I don't feel strongly about this though, if we feel people might misuse this. I've tidied up the PR, I think it's in relatively mergeable state, if we decide to proceed.

**Comment by maddinkunze** [comment]
Hi guys, i have another use-case for supporting the `Never` type in pydantic.

TL;DR: If the types accepted by an attribute are collected at runtime (i.e. not properly known beforehand), it may happen that there are no matching types (so `Never`)

I have written a helper that allows automatically collecting discriminated unions usually written like this

```python
class BaseLogger(BaseModel):
    def log(self, line: str): ...

class StdoutLogger(BaseLogger):
    type: Literal["stdout"]

class FileLogger(BaseLogger):
    type: Literal["file"]
    file: Path = Path("/tmp/log")

...

AnyLogger = Annotated[Union[StdoutLogger, FileLogger, ...], Field(discriminator="type")]
```

to a more convenient version

```python
class BaseLogger(PluginModel):
    def log(self, line: str): ...

class StdoutLogger(BaseLogger):
    type: Literal["stdout"]

class FileLogger(BaseLogger):
    type: Literal["file"]
    file: Path = Path("/tmp/log")

...

AnyLogger = PluginAdapter[BaseLogger]
```

e.g. for using with

```python
class AppConfig(BaseModel):
    logger: AnyLogger|None = None
```

It is primarily meant to allow libraries to define an interface where additional implementations can be added down the line

When writing this helper i hit the need for support of the `Never` type, because someone may define and (try to) use an interface that is never actually implemented:

```python
class BaseExtension(PluginModel):
    def run_hook(self): ...

AnyExtension = PluginAdapter[BaseExtension] # if there is no implementation of BaseExtension, this essentially reduces to Never

class Config(BaseModel):
    logger: AnyLogger|None = None
    extension: AnyExtension|None = None # so this is effectively Never|None and thus fails at schema creation time in pydantic
```

To solve this issue back then, the adapter type simply returns a custom `PydanticNever` type that looks pretty similar to the one in PR https://github.com/pydantic/pydantic/pull/13317; i really like the implementation proposed in the PR and would love to see it merged

In the end, this is similar to the generics use-case, where it is not known beforehand what types exactly are allowed. I am pretty sure there are other use-cases where the allowed types are collected dynamically, and in all of those cases it could be anything or nothing. But if it happens to be nothing, then it should be natively supported by pydantic

**Comment by maddinkunze** [comment]
By the way, the points described by @Viicos (in https://github.com/pydantic/pydantic/issues/9731#issuecomment-2194319078, treating `Never` in pydantic the same way as in python: the bottom type) also imply that `list[Never]` has the same _behaviour_ as `Annotated[list, MaxLen(0)]` (from https://github.com/pydantic/pydantic/issues/9731#issuecomment-2186620081, though a different _schema_), because `list[Never]` can never be anything but an empty list if i am not mistaken
